### DISCUSSION TITLE
Add BLE MIDI

### DISCUSSION THREAD

**Discussion by riban-bw** [InfiniTimeOrg/InfiniTime#1983] [Ideas]
This is likely to be a niche requirement but I plan to use my Pine Time as a MIDI controller and monitor. To this end it would be advantageous to add BLE MIDI to the stack. (This could be conditionally compiled if required by an app.) This could also be interfaced with the existing metronome to allow a MIDI message to drive the metronome (which could provide a solution for [this suggestion](https://github.com/InfiniTimeOrg/InfiniTime/discussions/1597)).

I would like to add BLE MIDI that advertises the BLE UUIDS:

- MIDI Service (UUID: 03B80E5A-EDE8-4B33-A751-6CE34EC4C700)
- MIDI Data I/O Characteristic (UUID: 7772E5DB-3868-4112-A1A9-F2669D106BF3)

It would decode incoming MIDI messages making them available as events and allow sending MIDI messages. I have already implemented a firmware for another wearable BLE device (TTGO Watch - [riband](https://github.com/riban-bw/riband) which I find very useful in controlling and monitoring my Zynthian (MIDI device). I wish to port this to InfinTime so that I may use it on my Pine Time.

As initially stated, this may be niche but the presence of a metronome in the core build suggests there are musicians using the InfiniTime so maybe this may be beneficial to more than a handful of people.

**Comment by riban-bw** [comment]
I have a proof of concept implementation running that advertises as BLE MIDI and can send MIDI events to a connected device. I have implemented the reception of MIDI events as a callback, i.e. BLE MIDI driver detects incoming MIDI messages and calls a function in an object that has registered for MIDI events. This is not a queue. Only one app may receive MIDI events. Is this right or should we have a queue of callbacks or use messaging? It looks like messages in RTOS are simple flags, i.e. do not carry data so a MIDI message would not hold the payload of the MIDI command and parameters. Have I missed something? Are we able to send RTOS messages with payload?

**Comment by sail2themoon** [comment]
Maybe this is somehow useful for you. Even though not the latest firmware: https://codeberg.org/Luno/InfiniTime_ble-midi
